src/Tasks/Common/FileUtilities.MetadataReader.cs

```diff
@@ -21,31 +22,60 @@ namespace Microsoft.NET.Build.Tasks
 {
     static partial class FileUtilities
     {
+        private static Dictionary<string, (DateTime LastKnownWriteTimeUtc, Version Version)> s_versionCache = new(StringComparer.OrdinalIgnoreCase /* Not strictly correct on *nix. Fix? */);
+
         private static Version GetAssemblyVersion(string sourcePath)
         {
-            using (var assemblyStream = new FileStream(sourcePath, FileMode.Open, FileAccess.Read, FileShare.Delete | FileShare.Read))
+            DateTime lastWriteTimeUtc = File.GetLastWriteTimeUtc(sourcePath);
+
+            if (s_versionCache.TryGetValue(sourcePath, out var cacheEntry) 
+                && lastWriteTimeUtc == cacheEntry.LastKnownWriteTimeUtc)
             {
-                Version result = null;
-                try
+                return cacheEntry.Version;
+            }
+
+            Version version = GetAssemblyVersionFromFile(sourcePath);
+
+            if (version is not null)
```

Comment:
Is there any reason not to cache a `null` return value for the version?  I think that means it wasn't a valid .NET assembly, but sometimes we do attempt to get the assembly version of those files, and it's not like the second time we try to do so it's going to be any more successful.

Reply:
Not that I can think of.